runtime/src/commonMain/kotlin/pbandk/UnknownField.kt

```diff
@@ -44,4 +44,101 @@ data class UnknownField internal constructor(val fieldNum: Int, val value: Value
+        is FieldDescriptor.Type.Primitive.Int32 -> (unknownField as UnknownField.Value.Varint).varint.toInt() as T
+        is FieldDescriptor.Type.Primitive.SInt32 -> (unknownField as UnknownField.Value.Varint).varint.toInt() as T //TODO
+        is FieldDescriptor.Type.Primitive.Int64 -> (unknownField as UnknownField.Value.Varint).varint as T
+        is FieldDescriptor.Type.Primitive.SInt64 -> (unknownField as UnknownField.Value.Varint).varint as T //TODO
+        is FieldDescriptor.Type.Primitive.UInt32 -> (unknownField as UnknownField.Value.Varint).varint.toInt() as T
+        is FieldDescriptor.Type.Primitive.UInt64 -> (unknownField as UnknownField.Value.Varint).varint as T
+    }
+}
+
+@Suppress("UNCHECKED_CAST")
+private fun <T : Message> parseUnknownFieldForTypeMessage(
+    type: FieldDescriptor.Type.Message<T>,
+    unknownField: UnknownField.Value
+): T {
+    return when (unknownField) {
+        is UnknownField.Value.Composite -> {
+            unknownField.values
+                .map { value ->
+                    type.messageCompanion.decodeFromByteArray((value as UnknownField.Value.LengthDelimited).bytes.array)
+                }
+                .reduce { acc, it -> acc.plus(it) as T }
+
+        }
+        is UnknownField.Value.LengthDelimited -> type.messageCompanion.decodeFromByteArray(unknownField.bytes.array)
+        else -> throw UnsupportedOperationException("unknown field is not supported: '${unknownField}'")
+    }
+}
+
+@Suppress("UNCHECKED_CAST")
+private fun <T : pbandk.Message.Enum> parseUnknownFieldForTypeEnum(
+    type: FieldDescriptor.Type.Enum<T>,
+    unknownField: UnknownField.Value
+): T {
+    return when (unknownField) {
+        is UnknownField.Value.Composite -> type.enumCompanion.fromValue((unknownField.values.last() as UnknownField.Value.Varint).varint.toInt())
+        is UnknownField.Value.LengthDelimited -> type.enumCompanion.fromValue((unknownField as UnknownField.Value.Varint).varint.toInt())
+        else -> throw UnsupportedOperationException("unknown field is not supported: '${unknownField}'")
+    }
+}
+
+@Suppress("UNCHECKED_CAST")
+private fun <K,V> parseUnknownFieldForTypeMap(
+    type: FieldDescriptor.Type.Map<K, V>,
+    unknownField: UnknownField.Value
+): MessageMap.Entry<K, V> {
+    return when (unknownField) {
+        is UnknownField.Value.Composite -> type.entryCompanion.decodeFromByteArray((unknownField.values.last() as UnknownField.Value.LengthDelimited).bytes.array)
+        is UnknownField.Value.LengthDelimited -> type.entryCompanion.decodeFromByteArray(unknownField.bytes.array)
+        else -> throw UnsupportedOperationException("unknown field is not supported: '${unknownField}'")
+    }
+}
+
+
+@Suppress("UNCHECKED_CAST")
+private fun <T : Any> parseUnknownFieldForTypeRepeated(
+    type: FieldDescriptor.Type.Repeated<T>,
+    unknownField: UnknownField.Value
+): Sequence<T> {
+    //TODO help
```

Comment:
@garyp I'm going to need some help here, not sure exactly what to do. I also feel like I did't do the `Map` type correctly either.